src/cocotb/_scheduler.py

```diff
@@ -341,170 +339,110 @@ def _test_completed(self, trigger=None):
-            is_first = True
-            self._pending_triggers.append(trigger)
-            while self._pending_triggers:
-                trigger = self._pending_triggers.pop(0)
-
-                if not is_first and isinstance(trigger, GPITrigger):
-                    self.log.warning(
-                        "A GPI trigger occurred after entering react - this "
-                        "should not happen."
-                    )
-                    assert False
+                self._state = GPIState.NORMAL
 
-                # this only exists to enable the warning above
-                is_first = False
+            self._react(trigger)
+            self._event_loop()
 
-                # When tasks run, they may append to our waiting list so the first
-                # thing to do is pop all tasks currently waiting on this trigger.
-                try:
-                    self._scheduling = self._trigger2tasks.pop(trigger)
-                except KeyError:
-                    # GPI triggers should only be ever pending if there is an
-                    # associated task waiting on that trigger, otherwise it would
-                    # have been unprimed already
-                    if isinstance(trigger, GPITrigger):
-                        self.log.critical(
-                            f"No tasks waiting on trigger that fired: {trigger}"
-                        )
-
-                        trigger.log.info("I'm the culprit")
-                    # For Python triggers this isn't actually an error - we might do
-                    # event.set() without knowing whether any tasks are actually
-                    # waiting on this event, for example
-                    elif _debug:
-                        self.log.debug(
-                            f"No tasks waiting on trigger that fired: {trigger}"
-                        )
+    def _react(self, trigger: Trigger) -> None:
+        """Called when a trigger fires."""
+        if _debug:
+            self.log.debug(f"Trigger fired: {trigger}")
 
-                    del trigger
-                    continue
+        # find all tasks waiting on trigger that fired
+        try:
+            scheduling = self._trigger2tasks.pop(trigger)
+        except KeyError:
+            # GPI triggers should only be ever pending if there is an
+            # associated task waiting on that trigger, otherwise it would
+            # have been unprimed already
+            if isinstance(trigger, GPITrigger):
+                self.log.critical(f"No tasks waiting on trigger that fired: {trigger}")
+                trigger.log.info("I'm the culprit")
+            # For Python triggers this isn't actually an error - we might do
+            # event.set() without knowing whether any tasks are actually
+            # waiting on this event, for example
+            elif _debug:
+                self.log.debug(f"No tasks waiting on trigger that fired: {trigger}")
```

Comment:
Doesn't need to be in this PR, but the codecov report indicates that we could add a test for this condition (since you're already describin the scenario in the comment above).

Reply:
I have thought about this. I'd like to add a test to the repo which runs `test_cocotb` with the scheduler debug on to hit these debug lines, then we can more accurately judge which cases we're actually missing.

Reply:
Filed #3968.